### PR TITLE
Refactored AbstractTreeViewer: extracted setRedraw

### PR DESCRIPTION
**Pull request by Wittmaxi**
Extracted from  #1072

### CODE REVIEW

File: bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java [hunk 1]
```diff
@@ -774,16 +774,15 @@ public void collapseAll() {
...
-			Widget w = internalGetWidgetToSelect(elementOrTreePath);
-			if (w != null) {
-				internalCollapseToLevel(w, level);
+		executeWithRedrawDisabled(new Runnable() {
+			@Override
+			public void run() {
+				Widget w = internalGetWidgetToSelect(elementOrTreePath);
+				if (w != null) {
+					internalCollapseToLevel(w, level);
+				}
```
Comment (HeikoKlare) [inline bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java]: You can simplify this by putting into a lambda expression:
```suggestion
		executeWithRedrawDisabled(() -> {
			Widget w = internalGetWidgetToSelect(elementOrTreePath);
			if (w != null) {
				internalCollapseToLevel(w, level);
			}
```
This also applies to the other instantiations of `Runnable`.

File: bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java [hunk 2]
```diff
@@ -3401,4 +3396,38 @@ public boolean contains(Object parent, Object element) {
 		return false;
 	}
 
+	/**
+	 * Transactionally perform an action on the TreeViewer without redrawing
+	 * regularly. Disables redraw, executes toExecute and enables redraw again.
```
Comment (HeikoKlare) [inline bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java]: nit: put `toExecute` in `@code{toExecute}` to make it easier to see that it's a code reference

File: bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java [hunk 3]
```diff
@@ -3401,4 +3396,38 @@ public boolean contains(Object parent, Object element) {
...
+			if (shouldDisableRedraw) {
+				control.setRedraw(false);
+			}
+			toExecute.run();
+		} finally {
+			if (shouldDisableRedraw) {
+				control.setRedraw(true);
+			}
+		}
+	}
```
Comment (HeikoKlare) [inline bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java]: This method is currently misleading, as its name states that it disables redraw while it accepts an argument that defines whether it actually does so.
I propose either of the folowing:
- Inline this code as it is only used once.
- Rename the method to something like `execute(Runnable toExecute, boolean withRedrawDisabled)`. In that case, the Javadoc has to be corrected as well (the second parameter is not documented).

**Comment by laeubi** [comment]
I think the wording "transactionally" is completely confusing here, a [transaction in computer sience](https://en.wikipedia.org/wiki/Transaction_processing)  something completely different than what happening here.

also `execute` is way to generic here, i would suggest using `runWithRedraw` or similar. Also note that this is not really safe to be called in a recursive manner as it does not restore the previous state, it simply assumes a state and in some cases reset it to this assumption. simple example I have previously already called `treeviewer.getControl().setRedraw(false)` this state will not be restored when I `labelProviderChanged` is called for whatever reason in between.

File: bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java [hunk 4]
```diff
@@ -1143,20 +1139,12 @@ public void expandToLevel(Object elementOrTreePath, int level) {
 	public void expandToLevel(Object elementOrTreePath, int level, boolean disableRedraw) {
 		if (checkBusy())
 			return;
-		Control control = getControl();
-		try {
-			if (disableRedraw) {
-				control.setRedraw(false);
-			}
+		execute(() -> {
```
Comment (laeubi) [inline bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java]: You should extract the runnable here and call run here directly if `disableRedraw = false` then you don't need the paramter what makes the whole thing way mor confusing then required.

**Comment by HeikoKlare** [comment]
> I think the wording "transactionally" is completely confusing here, a [transaction in computer sience](https://en.wikipedia.org/wiki/Transaction_processing) something completely different than what happening here.

👍 

> Also note that this is not really safe to be called in a recursive manner as it does not restore the previous state, it simply assumes a state and in some cases reset it to this assumption.

True. Even though this is existing behavior, it seems to be reasonable to store the initial redrawing state and reset this state after execution. The method should then document that in case the executed operation modifies the redraw state, this will be overwritten afterwards.

File: bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java (hunk 4 above)
Comment (HeikoKlare) [reply bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java]: I agree. That conforms to what I meant with the first option in https://github.com/eclipse-platform/eclipse.platform.ui/pull/1191#discussion_r1347120412 and I think it is actually the better of the two options.

**Comment by Wittmaxi** [comment]
@HeikoKlare there is currently no (... I couldn't find any ...) function that returns a redraw state in the public API of `Control`. The windows-Implementation has a method called `getDrawing` which isn't public API. The GTK-Implementation does not have that method. Hence, I see myself forced to invalidate the previous draw-state of the tree, unless you have a better idea?

At the very least, I can mark this with a comment in the JavaDoc, warning of invalidated states.
Since the tree is expected to "redraw by default" anyway, I don't think that it's too big of a stretch to assume the tree is always redrawing, except for operations that happen inside of `executeWithredrawDisabled` - maybe we can find a good solution for recursive functions that call `executeWithRedrawDisabled` themselves?

**Comment by HeikoKlare** [comment]
I see. When I take a look at all the callers of `setRedraw`, then most (all?) of them use is to temporarily disable redraw. I.e., the all follow the pattern of disabling redraw, doing something, and reenabling it afterwards. Many of them even don't re-enable redraw in a finally block to leave the system in a proper state in case something fails.

So I think this `executeWithoutRedraw` method should actually rather be placed inside `Control` and used by all these callers. To ensure that the method behaves as expected, it could ...
1. be  `synchronized` to avoid concurrent calls, and
2. maintain a number/stack of calls to ensure that redraw is only re-enabled when returning from the last call

Then we would have:
1. A reusable logic to execute something on a control while redrawing is disabled.
2. A fail-safe implementation of that logic, which always restores the old state, even if the code to be executed fails.
3. A thread-safe and recursion-aware implementation of this logic.

**Comment by iloveeclipse** [comment]
> 1. be  `synchronized` to avoid concurrent calls

SWT is single threaded, whoever calls redraw **must** be in UI thread. Please don't add synchronized in UI / SWT layer.

**Comment by HeikoKlare** [comment]
> > 1. be  `synchronized` to avoid concurrent calls
> 
> SWT is single threaded, whoever calls redraw **must** be in UI thread. Please don't add synchronized in UI / SWT layer.

Would probably be unnecessary anyway if the callers are tracked as proposed in 2. (which reduces to tracking the number of recursions in case of single-threaded access).

**Comment by vogella** [comment]
What is the status here?

**Comment by HeikoKlare** [comment]
This is to be supseded by a more general PR performing the refactoring at the level of `Controls` rather than on the `AbstractTreeViewer` only.

**Comment by vogella** [comment]
> supseded

So we can close this one? Do you have a link to the new PR?

**Comment by HeikoKlare** [comment]
From my point of view, we can close this one. A new PR does not exist yet. It will require some effort to adapt all places where redrawing is temporarily disabled, but I would prefer that over a local solution that only improves the behavior for one single kind of control. Maybe we can use the quiet period until 4.31 development opens to prepare an according PR.

**Comment by Wittmaxi** [comment]
Closed with regard to Heiko's comment, created #1285

**Pull request closed without merging**